src/types/types.zig

```diff
@@ -0,0 +1,376 @@
+//! This module defines the core data structures for representing types in the compiler's
+//! Hindley-Milner type inference system. It includes:
+//!
+//! - `Var`: unique type variable identifiers
+//! - `Descriptor`: the rank, mark, and structure of a type
+//! - `Content`: the semantic meaning of a type (flex var, alias, function, record, etc.)
+//! - `FlatType`: the 'flat' shape of a type (tuples, numbers, tag unions, etc.)
+//! - `Alias`: nominal or structural type aliases
+//! - `Func`, `Record`, `TagUnion`: structured type forms
+//!
+//! Special care is taken to keep memory layouts small and efficient. When modifying
+//! these types, please consider their size impact and unification performance.
+
+const std = @import("std");
+const collections = @import("../collections.zig");
+const Ident = @import("../base/Ident.zig");
+
+const MkSafeList = collections.SafeList;
+const MkSafeMultiList = collections.SafeMultiList;
+
+test {
+    // If your changes caused this number to go down, great! Please update it to the lower number.
+    // If it went up, please make sure your changes are absolutely required!
+    try std.testing.expectEqual(32, @sizeOf(Descriptor));
+    try std.testing.expectEqual(24, @sizeOf(Content));
+    try std.testing.expectEqual(16, @sizeOf(Alias));
+    try std.testing.expectEqual(20, @sizeOf(FlatType));
+    try std.testing.expectEqual(12, @sizeOf(Record));
+}
+
+/// A type variable
+pub const Var = enum(u32) {
+    _,
+
+    /// A safe list of type variables
+    pub const SafeList = MkSafeList(Var);
+};
+
+/// A type descriptor
+pub const Descriptor = struct { content: Content, rank: Rank, mark: Mark };
+
+/// A type variable rank
+pub const Rank = enum(u4) {
+    generalized = 0,
+    top_level = 1,
+    _,
+
+    /// Get the lowest rank
+    pub fn min(a: Rank, b: Rank) Rank {
+        return @enumFromInt(@min(@intFromEnum(a), @intFromEnum(b)));
+    }
+};
+
+/// A type variable mark
+pub const Mark = enum(u32) {
+    const Self = @This();
+
+    getVarNames = 0,
+    occurs = 1,
+    none = 2,
+    _,
+
+    /// Get the lowest rank
+    pub fn next(self: Self) Self {
+        return self + 1;
+    }
+};
+
+// content //
+
+/// Represents what the a type *is*
+///
+/// Numbers are special cased here. This means that when constraints, types
```

Comment:
```suggestion
/// Numbers are special-cased here. We represent them to the Roc programmer
/// as opaque types with phantom type variables, but since they come up
/// so often, we unify that representation into a special (much more
/// compact) representation which saves a lot of memory. For example, a type
```